### ISSUE TITLE
Support for Partially Expanding URI Templates

### ISSUE THREAD

**Issue by SJrX** [micronaut-projects/micronaut-core#4207]
Thanks for reporting an issue, please review the task list below before submitting the
issue. Your issue report will be closed if the issue is incomplete and the below tasks not completed.

NOTE: If you are unsure about something and the issue is more of a question a better place to ask questions is on Stack Overflow (https://stackoverflow.com/tags/micronaut) or Gitter (https://gitter.im/micronautfw/). DO NOT use the issue tracker to ask questions.

### Task List

- [x] Steps to reproduce provided
- [x] Stacktrace (if present) provided
- [x] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

1. Want to handle URI Templates in multiple places.
2. Use URITemplate class to do so.

### Expected Behaviour

A `partialExpand()` method that leaves variables that aren't provided unhandled.

### Actual Behaviour

Variables that aren't provided are removed.

### Environment Information

- **Operating System**: Al/l
- **Micronaut Version:** 2.0
- **JDK Version:** 14

### Example Application
Sorry that I didn't provide an example application, it wouldn't compile since I'm requesting Java API changes.

Here is another library that supports it:
- https://github.com/damnhandy/Handy-URI-Templates/blob/master/src/main/java/com/damnhandy/uri/template/UriTemplate.java#L386

### Background

The main motivation for this is if I have some endpoint `POST /foo/{foo-id}/bar/{bar-id}/zoo/` that creates a resource and generates a link for: `/foo/{foo-id}/bar/{bar-id}/zoo/{zoo-id}`, and I need to generate a link for in the response. The most natural place in the code for me to generate the link doesn't have `foo-id` or `bar-id`. However in almost every case, we could fill in these variables from the endpoint that we are calling and filling them in with that context. So I'd like to fill in: 
`/foo/{foo-id}/bar/{bar-id}/zoo/{zoo-id}` with the zoo-id in one place and the rest in a filter.

**Comment by yawkat** [comment]
I am against this because this pattern can create a template injection vulnerability.

- pattern is `/foo/{foo}/bar/{bar}`
- server rewrites `{foo}` with user-provided data
- user provides `{some-secret}` as the data
- url is rewritten to `/foo/{some-secret}/bar/{bar}`
- the endpoint does the remaining substitutions, and would inject `some-secret` as well as `bar`, even though the developer only intended to inject `bar`.

To put it in more general terms: After performing one template substitution, the output is often [tainted](https://en.wikipedia.org/wiki/Taint_checking). Doing a second template substitution on the output can be dangerous.

We can't prevent people from using this pattern ofc, but I don't think we should encourage this with a feature that can be used for little else.

**Comment by SJrX** [comment]
Admittedly it has been about 16 months since I filed this issue, but I'm not sure I agree with your objections (but haven't looked at the code closely) @yawkat . 

> I am against this because this pattern can create a template injection vulnerability.
...
> user provides {some-secret} as the data
> url is rewritten to /foo/{some-secret}/bar/{bar}
> the endpoint does the remaining substitutions, and would inject some-secret as well as bar, even though the developer only intended to inject bar.

My two main objections are that:

1. I assume that URI templates are not using random substitution of any value, but of particular values passed in via a Map or Bean, it seems odd that that map would have these secrets in them to fill in Uri templates.
2. I would just assume that you shouldn't be able to substitute non URL safe characters as part of the template value, as you don't get a valid URL out.

**Comment by yawkat** [comment]
Yes, if the input map is limited this is a non-issue, but I'd not depend on it to be. And who knows what downstream templating engines will do (since they probably won't be another micronaut instance), they might be vulnerable to more severe attacks, like format string attacks. Just look at what happened with log4j when they tried to apply templating on tainted data.

Encoding should protect against this yes, but not all substitutions are encoded, e.g. fragment substitutions (`https://example.com/{#foo}`)

**Comment by graemerocher** [comment]
Triaged as Feature with a security concern. `UriTemplate.expand` currently drops variables that are not present; partially expanding and then re-expanding the output can allow user-controlled values to introduce new template expressions. A safer design is a typed/opaque partial template or an API that returns remaining variables without reparsing substituted values. If a string API is added, encode substitution values and explicitly reject nested expansion and fragment operators. Add staged-expansion and injection-payload tests before considering this for Core.
